Upgrade clang-tidy used in CI to LLVM 16

jbytheway opened the pull request:
#### Summary
Infrastructure "Upgrade clang-tidy used in CI to LLVM 16"

#### Purpose of change
(This work started by @BrettDong in #64699; I'm picking this up to try and push this over the finish line)

We have custom `clang-tidy` checks, but the `clang-tidy` version we are using does not support plugins out-of-the-box. So we have to use a [patched version of clang-tidy](https://github.com/jbytheway/clang-tidy-plugin-support) to run our custom checks. This makes setting up clang-tidy checks in local development environments a little bit involved, especially for non-Debian/Ubuntu distro users.

Fortunately, [plugin support](https://reviews.llvm.org/D111100) landed in a later LLVM release. This simplifies `clang-tidy` workflow a lot. After building our custom checks as a shared library, we can just run the checks with a vanilla version of `clang-tidy`.

#### Describe the solution
* Update the custom check code to be compatible with new LLVM.
* Need to ship our own copy of `check_clang_tidy.py` since it's no longer coming from the patched `clang-tidy`.
* Simplify the CMake code needed to support the old approach.
* Disable all the checks that trigger failures after this change.
* Change a particular array to a `std::array` to fix a mystery crash (thanks to @andrei8l) for figuring that out).

#### Describe alternatives you've considered
We could try to fix the warnings before updating LLVM, but that requires a lot of potential compromises and difficult decisions, and risks actual code breakage, which I'd rather not mingle with this change.

#### Testing
Run clang-tidy.

#### Additional context

BrettDong committed:
Update custom clang-tidy checks to be compatible with LLVM 16

BrettDong committed:
Try clang-tidy plugin with LLVM 16 on CI

BrettDong committed:
Exclude check_clang_tidy.py from flake8 check

BrettDong committed:
Update cata-almost-never-auto check

BrettDong committed:
Add tools/ to astyle-diff Makefile target

jbytheway committed:
Remove a C-style array

This array causes clang-tidy to crash.  No idea why.

jbytheway committed:
Port clang-tidy helper script to LLVM 16

Needs to conform to new clang-tidy CLI.

jbytheway committed:
Disable all checks which fire on LLVM 16

We don't want to have to fix all the new issues that LLVM 16 finds up
front, so just disable the checks for now.  Will be re-enabled slowly
over time.

Qrox commented on tools/clang-tidy-plugin/test/lit.cfg:
```diff
@@ -10,11 +10,8 @@ config.test_exec_root = os.path.join(
-    check_clang_tidy = os.path.join(
-            config.cata_source_dir, "./tools/clang-tidy-plugin/test/check_clang_tidy.py" )
+check_clang_tidy = os.path.join(
+        config.cata_source_dir, "./tools/clang-tidy-plugin/test/check_clang_tidy.py" )
```
This is needed on Windows to add custom executable path argument to the `check_clang_tidy.py` call.

Qrox commented on tools/clang-tidy-plugin/test/lit.site.cfg.in:
```diff
@@ -2,7 +2,6 @@ import os
 
 config.cata_source_dir = "@CMAKE_SOURCE_DIR@"
 config.plugin_build_root = "@CMAKE_CURRENT_BINARY_DIR@"
-config.cata_check_clang_tidy = "@CATA_CHECK_CLANG_TIDY@"
```
This is needed on Windows to add custom executable path argument to the `check_clang_tidy.py` call.

jbytheway replied:
Are you saying that on Windows you need a different `check_clang_tidy.py`, or that you need to pass different arguments to `check_clang_tidy.py`?

Qrox replied:
Both at the moment. I think I can switch to using `check_clang_tidy.py` in this PR but even then a different argument is needed, and I'll need to add my patches for it to work.

jbytheway replied:
I have mixed feelings about having custom patches to the in-repo `check_clang_tidy.py`.  On the one hand it's convenient.  On the other it will make it annoying to incorporate upstream changes to the file.

For now I guess I'll put these features back in so they're available to Windows folks if they need them.

Qrox commented on CMakeLists.txt:
```diff
@@ -29,9 +29,6 @@ option(CATA_CCACHE "Try to find and build with ccache" "ON")
 option(CATA_CLANG_TIDY_PLUGIN "Build Cata's custom clang-tidy checks as a plugin" "OFF")
 option(CATA_CLANG_TIDY_EXECUTABLE "Build Cata's custom clang-tidy checks as an executable" "OFF")
 option(TESTS "Compile Cata's tests" "ON")
-set(CATA_CLANG_TIDY_INCLUDE_DIR "" CACHE STRING
-        "Path to internal clang-tidy headers required for plugin (e.g. ClangTidy.h)")
-set(CATA_CHECK_CLANG_TIDY "" CACHE STRING "Path to check_clang_tidy.py for plugin tests")
```
This is required for compiling clang-tidy on Windows.

Qrox commented on tools/clang-tidy-plugin/CMakeLists.txt:
```diff
@@ -58,10 +58,6 @@ endif ()
-if (NOT "${CATA_CLANG_TIDY_INCLUDE_DIR}" STREQUAL "")
-    target_include_directories(${CataAnalyzerName} SYSTEM PRIVATE ${CATA_CLANG_TIDY_INCLUDE_DIR})
-endif ()
-
```
This is required for compiling clang-tidy on Windows.

jbytheway replied:
Do you think it makes sense to still have this specific feature rather than just allowing folks to pass an extra include dir via `CMAKE_CXX_FLAGS`?

Qrox replied:
I think `CMAKE_CXX_FLAGS` applies to all targets though. Is there a way to specify include dir for specific targets through the command line?

jbytheway replied:
No, I don't think so.  So indeed it's probably sensible to keep this (which I think is what I did end up doing in this PR).

Qrox commented on tools/clang-tidy-plugin/AlmostNeverAutoCheck.cpp:
```diff
@@ -91,6 +91,7 @@ static void CheckDecl( AlmostNeverAutoCheck &Check,
     }
 
     QualType AutoTp = Initializer->getType();
+    QualType DesugaredAutoTp = AutoTp.getDesugaredType( *Result.Context );
```
```suggestion
```

Qrox commented on tools/clang-tidy-plugin/AlmostNeverAutoCheck.cpp:
```diff
@@ -107,6 +108,7 @@ static void CheckDecl( AlmostNeverAutoCheck &Check,
     PrintingPolicy Policy( LangOptions{} );
     Policy.adjustForCPlusPlus();
     std::string TypeStr = AutoTp.getAsString( Policy );
+    const std::string DesugaredTypeStr = DesugaredAutoTp.getAsString( Policy );
```
```suggestion

    std::string DesugaredTypeStr;
    // Test stripped type is not null to avoid a crash in
    // QualType::getSplitDesugaredType in clang/lib/AST/Type.cpp
    if( QualifierCollector().strip( AutoTp ) ) {
        QualType DesugaredAutoTp = AutoTp.getDesugaredType( *Result.Context );
        DesugaredTypeStr = DesugaredAutoTp.getAsString( Policy );
    }
```

There was a crash here in the original PR and this seems to fix it.

Qrox commented on tools/clang-tidy-plugin/CMakeLists.txt:
```diff
@@ -58,26 +58,6 @@ endif ()
-    target_include_directories(${CataAnalyzerName} SYSTEM PRIVATE ${CATA_CLANG_TIDY_INCLUDE_DIR})
-endif ()
-
 target_compile_definitions(${CataAnalyzerName} PRIVATE ${LLVM_DEFINITIONS})
```
```suggestion
separate_arguments(LLVM_DEFINITIONS_SEP UNIX_COMMAND "${LLVM_DEFINITIONS}")
target_compile_definitions(${CataAnalyzerName} PRIVATE ${LLVM_DEFINITIONS_SEP})
```

This is needed to correctly separate LLVM compile definitions on Windows. (Otherwise all definitions are passed to the compiler as a single quoted string.)

Qrox commented:
The include paths in UseStringViewCheck.h/cpp need to be updated too.

jbytheway committed:
Update LLVM version in build scripts

jbytheway committed:
Support Windows arg parsing in plugin CMakeLists

Co-authored-by: Jianxiang Wang (王健翔) <qrox@sina.com>

jbytheway committed:
Work around crash in AlmostNeverAuto

Trying to desugar certain types would crash inside the Clang code; this
works around that issue.

Co-authored-by: Jianxiang Wang (王健翔) <qrox@sina.com>

jbytheway commented:
OK, I think this is ready.

@Qrox I imagine there will probably be more tweaks needed for your workflow to work on Windows, but it's probably simpler for you to work on that in a separate PR.

Qrox commented:
Well, I've already updated my workflow on Windows and the only changes needed seem to be updating the llvm repo and the include dir in the CMake command. The documentation needs to be updated too but that can be done in a separate PR.
